### ISSUE TITLE
Implicit variable projectDir breaks caching

### ISSUE THREAD

**Issue by grst** [nextflow-io/nextflow#2186]
### Expected behavior and actual behavior

Using the `$projectDir` variable in the `script:` section breaks caching (the process gets executed even though nothing has changed). Interestingly, using `$baseDir` (which is deprecated) behaves as expected. 

### Steps to reproduce the problem

```nextflow
#!/usr/bin/env nextflow

nextflow.enable.dsl = 2

process test_cache_projectDir {
    output:
        path("result.txt"), emit: result
    
    script:
    """
    file ${projectDir}/foo.txt > result.txt
    """
}

process test_cache_baseDir {
    output:
        path("result.txt"), emit: result
    
    script:
    """
    file ${baseDir}/foo.txt > result.txt
    """
}

workflow {
    test_cache_projectDir()
    test_cache_baseDir()
}
```

### Program output 

First run
```console
> nextflow run main.nf 
N E X T F L O W  ~  version 21.04.0
Launching `main.nf` [friendly_dubinsky] - revision: 9d548112c1
executor >  local (2)
[4f/4568da] process > test_cache_projectDir [100%] 1 of 1 ✔
[90/fb827c] process > test_cache_baseDir    [100%] 1 of 1 ✔
```

Second run
```console
> nextflow run main.nf -resume
N E X T F L O W  ~  version 21.04.0
Launching `main.nf` [fervent_meucci] - revision: 9d548112c1
executor >  local (1)
[5d/7f5c4d] process > test_cache_projectDir [100%] 1 of 1 ✔
[90/fb827c] process > test_cache_baseDir    [100%] 1 of 1, cached: 1 ✔
```

[.nextflow.log.1.txt](https://github.com/nextflow-io/nextflow/files/6711391/default.nextflow.log.1.txt)
[.nextflow.log.txt](https://github.com/nextflow-io/nextflow/files/6711392/default.nextflow.log.txt)

### Environment 

* Nextflow version: `version 21.04.0 build 5552`
* Java version: `openjdk version "11.0.8-internal" 2020-07-14`
* Operating system: arch linux
* Bash version: `GNU bash, version 5.1.8(1)-release (x86_64-pc-linux-gnu)`

### Additional context

I use `$projectDir` to access assets shipped with the pipeline (e.g. a report template)

CC @riederd

**Comment by pditommaso** [comment]
Well spotted! Patched, it will be included in next release

**Comment by pditommaso** [comment]
Included in release `21.06.0-edge`.
